File: 144_Unique Paths II.cpp
/*
144. Unique Paths II
Number of unique paths in grid with obstacles (1 = obstacle). Use DP in-place or 1D array.
*/
#include <bits/stdc++.h>
using namespace std;

int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
    int m = obstacleGrid.size(), n = obstacleGrid[0].size();
    vector<int> dp(n,0);
    dp[0] = obstacleGrid[0][0] == 0 ? 1 : 0;
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            if (obstacleGrid[i][j] == 1) dp[j] = 0;
            else if (j > 0) dp[j] += dp[j-1];
        }
    }
    return dp[n-1];
}

int main() {
    vector<vector<int>> g{{0,0,0},{0,1,0},{0,0,0}};
    cout << "uniquePathsWithObstacles = " << uniquePathsWithObstacles(g) << "\n"; // 2
    vector<vector<int>> g2{{0,1},{0,0}};
    cout << "uniquePathsWithObstacles = " << uniquePathsWithObstacles(g2) << "\n"; // 1
    return 0;
}

// Time Complexity: O(m*n)
// Space Complexity: O(n)

File: 144_Unique Paths II.py
"""
144. Unique Paths II
DP with obstacles using 1D rolling array.
"""
from typing import List

def unique_paths_with_obstacles(obstacleGrid: List[List[int]]) -> int:
    m, n = len(obstacleGrid), len(obstacleGrid[0])
    dp = [0] * n
    dp[0] = 1 if obstacleGrid[0][0] == 0 else 0
    for i in range(m):
        for j in range(n):
            if obstacleGrid[i][j] == 1:
                dp[j] = 0
            elif j > 0:
                dp[j] += dp[j-1]
    return dp[-1]

if __name__ == '__main__':
    print('uniquePathsWithObstacles =', unique_paths_with_obstacles([[0,0,0],[0,1,0],[0,0,0]]))
    print('uniquePathsWithObstacles =', unique_paths_with_obstacles([[0,1],[0,0]]))

# Time Complexity: O(m*n)
# Space Complexity: O(n)
